### ISSUE TITLE
Templates identifiers called as functions must include at least one visible character.

### ISSUE THREAD

**Issue by jagracey** [eslint/eslint#6417]
#### **Proposal:** Templates called as functions must include at least one visible character.
<br>
**Context:**
Unicode and ES6 support the multiple zero-width characters as variable names (both ID_Start & ID_Continue). Newer editors like Atom don't by default display zero-width characters while other editors like Vim display by default many zero-width characters as Code Point values, but many zero-
width chars still sneak by, like Variation Selectors (U+E0100 - U+E01EF).

<br>
**When does this rule warn? Please describe and show example code:**

:-1: Example of incorrect code for this rule:
```javascript
var \u{200AD}= evilFunc;  // A single line can be all it takes to subtly call code.

// This template is preceded by a zero-width char,
// executing evilFunc with 'Bar' as a parameter.
𠂭`Bar`
```

:+1: Example of correct code for this rule:
```javascript
foo𠂭`Bar`  // Still includes an zero-width char, but includes a visible char, so is okay.

`Bar`     // Does not include a zero-width char, so is okay.
```

<br>
**Is this rule preventing an error or is it stylistic?**
Security rule.

<br>
**Why is this rule a candidate for inclusion instead of creating a custom rule?**
It is widely applicable to ES6 users, and can be considered fairly important. The code that developers read must behave predictably.

<br>
**Are you willing to create the rule yourself?**
Yes.

**Comment by platinumazure** [comment]
I'm a bit confused by the title-- you say "templates called as functions" but this seems to be about identifiers. Can you please clarify?

**Comment by mysticatea** [comment]
Is [no-irregular-whitespace] not enough?

[no-irregular-whitespace]: http://eslint.org/docs/rules/no-irregular-whitespace

**Comment by jagracey** [comment]
@platinumazure  Yes, the title was a bit confusing. To clarify, the identifier (variable name) of a "tagged template literal" should possess at least one visible character.

**Comment by michaelficarra** [comment]
@mysticatea No, it's not whitespace.

@jagracey Only when the tag is an identifier, of course. But I don't see any reason to limit this to template tags; grouping parentheses can also lead to covertly calling or constructing a function. You should just have it apply to all identifiers.

**Comment by jagracey** [comment]
I recently found out about U+3614, most of the time it shows up as a whitespace character, but has the ID_Start property.  I made a [post](https://www.reddit.com/r/programming/comments/4o7qbw/variable_identifiers_can_effectively_include/) with more details.

Given this finding, this issue with zero-width characters is superseded by whitespace characters in variable names.

**Comment by alberto** [comment]
Thanks for your interest in improving eslint. Unfortunately, it looks like consensus couldn't be reached on this issue and so I'm closing it. While we wish we'd be able to accommodate everyone's requests, we do need to prioritize. We've found that issues failing to reach consensus after 21 days tend never to reach consensus, and as such, we close those issues. This doesn't mean the idea isn't interesting, just that it's not something the team can commit to.
